xarray/backends/rasterio_.py

```diff
@@ -0,0 +1,131 @@
+import numpy as np
+
+try:
+    import rasterio
+except ImportError:
+    rasterio = False
+
+from .. import Variable
+from ..core.utils import FrozenOrderedDict, Frozen, NDArrayMixin, is_scalar
+from ..core import indexing
+from ..core.pycompat import OrderedDict, suppress
+
+from .common import AbstractDataStore
+
+_rio_varname = 'raster'
+
+_error_mess = ('The kind of indexing operation you are trying to do is not '
+               'valid on RasterIO files. Try to load your data with ds.load()'
+               'first.')
+
+
+class RasterioArrayWrapper(NDArrayMixin):
+    def __init__(self, ds):
+        self.ds = ds
+        self._shape = self.ds.count, self.ds.height, self.ds.width
+        self._ndims = len(self.shape)
+
+    @property
+    def dtype(self):
+        return np.dtype(self.ds.dtypes[0])
+
+    @property
+    def shape(self):
+        return self._shape
+
+    def __getitem__(self, key):
+
+        # make our job a bit easier
+        key = indexing.canonicalize_indexer(key, self._ndims)
+
+        # bands cannot be windowed but they can be listed
+        bands, n = key[0], self.shape[0]
+        if isinstance(bands, slice):
+            start = bands.start if bands.start is not None else 0
+            stop = bands.stop if bands.stop is not None else n
+            if bands.step is not None and bands.step != 1:
+                raise IndexError(_error_mess)
+            bands = np.arange(start, stop)
+        # be sure we give out a list
+        bands = (np.asarray(bands) + 1).tolist()
+
+        # but other dims can only be windowed
+        window = []
+        squeeze_axis = []
+        for i, (k, n) in enumerate(zip(key[1:], self.shape[1:])):
+            if isinstance(k, slice):
+                start = k.start if k.start is not None else 0
+                stop = k.stop if k.stop is not None else n
+                if k.step is not None and k.step != 1:
+                    raise IndexError(_error_mess)
+            else:
+                if is_scalar(k):
+                    # windowed operations will always return an array
+                    # we will have to squeeze it later
+                    squeeze_axis.append(i+1)
+                k = np.asarray(k).flatten()
```

Comment:
For clarity, I would make the scalar case separate and not use `.flatten()` (which could have other undesirable consequences)